Clear client data when an account switched

MisRob opened the pull request:
## Description

Fix bugs caused by IndexedDB data being persisted when switching between more user accounts.

#### Issue Addressed

Fixes #2062 and following Notion issues:
- https://www.notion.so/learningequality/Logging-out-with-Email-A-then-logging-in-with-Email-B-shows-the-channels-from-Email-A-in-the-My-Cha-7adc1e1f4756466abbe63d016a7c0bfe
- https://www.notion.so/learningequality/My-collections-made-as-user-b-com-showed-up-when-I-log-in-as-a-a-com-because-the-information-remains-153dec110580467890b01907f61c01dc
- https://www.notion.so/learningequality/Created-another-new-account-clicking-on-the-validation-link-takes-me-to-my-other-existing-Studio-a-2b0eae881441420b9d61f860a68b5fcb
- https://www.notion.so/learningequality/For-some-reason-when-i-clicked-on-the-activation-link-sent-to-my-yahoo-email-it-had-a-channel-that-f638cdb9b1c240d395e710ec700cbfcd
- https://www.notion.so/learningequality/Seems-weird-that-I-can-see-other-users-collections-just-by-going-to-the-collections-tab-856ac2fdfc5b4146af55d94b03d6f100

## Steps to Test

- [ ] *Delete "KolibriStudio" IndexedDB because there are updates to its schema*
- [ ] *Check that data (collections, channels) of one user are not displayed when signed in as another user*
- [ ] *Check that when you log out in one tab, you will be logged out in another*

## Implementation Notes

I've added a new session table to IndexedDB where the current user object is saved and connected `session` Vuex store to it. When starting an app, the current user id is compared to `window.user.id` (or `null` for guests). If those values are different, the whole IndexedDB will be cleared.

#### Does this introduce any tech-debt items?

I am going to push some tests here as soon as @rtibbles finishes WIP PR review and I am sure that this is the right direction.

## Checklist

- [ ] Is the code clean and well-commented?
- [ ] Has the `docs` label been added if this introduces a change that needs to be updated in the [user docs](https://kolibri-studio.readthedocs.io/en/latest/index.html)?
- [ ] Has the `CHANGELOG` label been added to this pull request? Items with this label will be added to the [CHANGELOG](https://github.com/learningequality/studio/blob/master/CHANGELOG.md) at a later time
- [ ] Are there tests for this change?
- [ ] Are all user-facing strings translated properly (if applicable)?
- [ ] Has the `notranslate` class been added to elements that shouldn't be translated by Google Chrome's automatic translation feature (e.g. icons, user-generated text)?
- [ ] Are all UI components LTR and RTL compliant (if applicable)?
- [ ] Are views organized into `pages`, `components`, and `layouts` directories [as described in the docs](https://github.com/learningequality/studio/blob/vue-refactor/docs/architecture.md#where-does-the-frontend-code-live)?
- [ ] Are there any new ways this uses user data that needs to be factored into our [Privacy Policy](https://github.com/learningequality/studio/tree/master/contentcuration/contentcuration/templates/policies/text)?
- [ ] Are there any new interactions that need to be added to the [QA Sheet](https://docs.google.com/spreadsheets/d/1HF4Gy6rb_BLbZoNkZEWZonKFBqPyVEiQq4Ve6XgIYmQ/edit#gid=0)?
- [ ] Are there opportunities for using Google Analytics here (if applicable)?
- [ ] If the Pipfile has been changed, is the updated Pipfile.lock file also included in this PR?
- [ ] Are the migrations [safe for a large db](https://www.braintreepayments.com/blog/safe-operations-for-high-volume-postgresql/) (if applicable)?

MisRob commented on contentcuration/contentcuration/frontend/shared/app.js:
```diff
@@ -43,20 +44,31 @@ Vue.component('ActionLink', ActionLink);
+
+  const userId = window.user ? window.user.id : null;
+  const isAnonymous = userId === null;
+
```
I am not sure how to handle errors for database Promises on lines 53 through 60. I was thinking about calling `resetDB()` whenever anything goes wrong here and logging the error through Raven but I am not sure if it's worth it, these calls don't look so problematic. What's your general strategy?

MisRob commented on contentcuration/contentcuration/frontend/shared/data/index.js:
```diff
@@ -35,6 +36,9 @@ export function setupSchema() {
     [CHANGE_LOCKS_TABLE]: 'id++,tracker_id,expiry',
+    // A special table for keeping a current user session
+    // (it contains only a single row)
+    [CURRENT_SESSION_TABLE]: '',
```
I assume that there is no need to increase the database version on schema updates while still on `develop`, right?

rtibbles replied:
No, not currently - and unfortunately, we would need to add some additional logic to handle version upgrades.

rtibbles commented:
> That makes me think whether it would make sense to rather connect this module to IndexedDB through Resources

I should have been more explicit in my comment on #2062 - this was what I was thinking, and that we should use the listeners to set the Vuex state. This would be useful when logout happens in one tab, we can then propagate the logout to all open tabs in the browser.

MisRob commented:
Ah okay, that's fine, I misunderstood.

> This would be useful when logout happens in one tab, we can then propagate the logout to all open tabs in the browser.

This sounds good. Thank you, I am going to update the PR.

MisRob commented on contentcuration/contentcuration/frontend/shared/data/resources.js:
```diff
@@ -536,6 +536,14 @@ class IndexedDBResource {
+    return this.table.toCollection().first();
+  }
+
+  indexNotEqual(index, value) {
```
I added this function because it seems that it's not possible to access `notEqual` directly on IndexedDBResource after calling `where` on it. I am thinking whether it would make sense to somehow expose Dexie methods so we can use its API instead of adding more and more custom methods on our wrapper like here. Or is it possible to use it somehow now?

rtibbles replied:
Can always just call `SessionResource.table.where(index).notEqual(value)` from wherever you are invoking it.

MisRob replied:
Thank you, accessing the table directly in this way makes sense. I am also going to remove the `first` method from `IndexedDBResource`.

MisRob replied:
We'll just need to be careful to know the difference between using `Resource.table.where` and our own implementation for `Resource.where`

MisRob commented on contentcuration/contentcuration/frontend/shared/vuex/session/index.js:
```diff
@@ -85,20 +100,28 @@ export default {
+      // though, because dexie-observable delete listener handler would
+      // be run and depending on timing, it might remove the entry after
+      // it's added and the current user would be lost)
+      await Session.indexNotEqual('id', currentUser.id).delete();
```
I think I might solve this in a much better way, and that would be to rather always use a constant key for a current user instead of their id in the session table. That way, we would be really sure that there is only one entry, and I would feel confident removing this table reset completely. It seems that I would need to adjust IndexedDBResource `put` to accept an optional key [as Dexie does](https://dexie.org/docs/Table/Table.put()#syntax). I am not yet confident enough to update our wrapper API, so at first, I wanted to ask. Do you foresee any obstacles or does it sound like a good idea?

rtibbles replied:
See my suggestion above - by changing the default id field to something other than `id` we can use that and just set it to `CURRENT` or some other such constant, and we only ever put that user into the DB.

`await Session.delete('CURRENT');` would then achieve the same thing here.

MisRob replied:
That might still cause the problem described in the comment above this line.

MisRob replied:
From now on, we'll ensure having exactly one entry using the constant ID though so I think it's safe to remove this completely.

MisRob commented:
@rtibbles I've updated the PR, its description and added some new questions I'd appreciate your opinion about. Thank you.

rtibbles commented on contentcuration/contentcuration/frontend/shared/data/resources.js:
```diff
@@ -677,6 +685,11 @@ class Resource extends mix(APIResource, IndexedDBResource) {
 
+export const Session = new IndexedDBResource({
+  tableName: TABLE_NAMES.SESSION,
+  uuid: false,
```
To address your suggestion below, can add:
```
idField: 'constantId',
```
here - and then just query by the constant value you set for users when you set to the table.

rtibbles commented on contentcuration/contentcuration/frontend/shared/vuex/session/index.js:
```diff
@@ -54,23 +56,36 @@ export default {
+      // prevent from double reloading of the login page
+      // ('REMOVE_SESSION' might be triggered by dexie-observable
+      // when reseting IndexedDB)
+      const ACCOUNTS_APP_URL = '/accounts/';
```
We should reference `window.Urls.accounts()` for the URL here.

rtibbles commented on contentcuration/contentcuration/frontend/shared/vuex/session/index.js:
```diff
@@ -85,20 +100,28 @@ export default {
+      // it's added and the current user would be lost)
+      await Session.indexNotEqual('id', currentUser.id).delete();
+
+      await Session.put(currentUser);
```
Would then need to make sure that we add `constantId: 'CURRENT'` to the `currentUser` object before we do the put here.

rtibbles reviewed (commented):
Some suggestions for how to simplify based on your question.

MisRob commented:
@rtibbles Thank you. I've just addressed all issues except of the constant index. Regarding that:

1. When I asked about that, I was rather proposing using [a solution similar to what I played with](https://github.com/learningequality/studio/pull/2490/commits/0872ba3755656da58b90241afaf65f1ee159d14c#diff-f5a688eac83b9eaeb8320505c4528603197241c5da09b92d916c5d0ff0724d9cR58) in the first (now reverted) version (except that there would be a whole user object instead of just `userId`) using [Dexie's hidden primary key](https://dexie.org/docs/Version/Version.stores()#detailed-schema-syntax). For that purpose, I could use its native API again. But I am hesitant to do so when updating data. I'd prefer using only our Resource methods for that. So I was rather wondering if it would make sense to support it in our Resources. Although this seems to be ideal for this particular use-case, I am not sure about the complexity (I guess other methods would need to support hidden indices in that case?) and whether it's worth it at this point. Or is there a simple way how to achieve this right now on our wrapper that I missed? In any case, it's rather nice-to-have.

2. When it comes to your suggestion to save the constant to the user object, I am not sure about its advantage when compared to the current approach. With both approaches, we still need to be careful to take some preliminary steps before saving a new entry. Clearing other possible entries seems a bit more intuitive than updating the current user object with a new field so I am fine with it as it is now. When it comes to these two solutions, I don't have a strong preference though.

rtibbles commented:
I am not sure I see an advantage to using a hidden primary key - I think we should either just always clear the table before adding  new entry and use the user id as the primary key, or do what I suggest with the constant id.

The reason for using the constant ID is that it will then work seamlessly with updates based using our Vuex listeners - so when our mutations for the session state trigger, we can ignore anything that doesn't have that constant id (in case something weird is happening), and we are mostly just looking out for create and update events, because the user would never get cleared, it either gets set to a new user or to an anonymous user.

MisRob commented:
Thank you. Okay, so I will leave the logic related to clearing the table before saving a new user and will also use the constant ID and try to add some tests. That should be hopefully all needed to finish this (I'd like to move user preferences too though would prefer doing that in the next PR if it's fine. I've started working on validation meanwhile, this fixes the referenced issues and has been open a while.

MisRob commented:
> I am not sure I see an advantage to using a hidden primary key

I just thought that it would be simpler not to need to add some extra property to the user object every time we work with it. It also seems a bit more readable (I can imagine that the purpose of the constant field might not be apparent to folks who are not familiar with the specific reason for adding it). But that's basically all. I'll go with the constant field for now.

rtibbles commented on contentcuration/contentcuration/frontend/shared/vuex/session/index.js:
```diff
@@ -85,20 +99,23 @@ export default {
-        context.commit('SET_CURRENT_USER', {});
-        localStorage['loggedOut'] = true;
-        window.location = '/';
+        context.commit('REMOVE_SESSION');
```
Is there a benefit to not clearing the indexedDB explicitly here when we know a logout has occurred? Combined with the suggestion I made above for the the conditional, this should limit the number of times we end up clearing the database.

MisRob replied:
I think that I've done it this way because of a different set of conditions in `startApp` where logout was handled. I didn't have a particular reason for that, though. Your suggestion makes sense. It saves one reset call :+1:

rtibbles commented on contentcuration/contentcuration/frontend/shared/app.js:
```diff
@@ -43,20 +45,34 @@ Vue.component('ActionLink', ActionLink);
+  const dbCurrentUser = (await Session.get(CURRENT_USER)) || {};
+
+  if (
+    currentUser.id === undefined ||
```
I think we can simplify this conditional? The main thing is making sure if the previously logged in user was not a guest, and now the current user is a different user?

```
dbCurrentUser.id && dbCurrentUser.id !== currentUser.id
```

rtibbles commented on contentcuration/contentcuration/frontend/shared/vuex/session/index.js:
```diff
@@ -54,23 +57,34 @@ export default {
+      // prevent from double reloading of the login page
+      // ('REMOVE_SESSION' might be triggered by dexie-observable
+      // when reseting IndexedDB)
+      if (!window.location.pathname.endsWith(window.Urls.accounts())) {
```
This feels outside of the scope of a Vuex mutation - presumably the goal here is to redirect whenever there is a change to the current user, but I don't think we should do it inside of mutation.

I would suggest either adding a vuex watcher to handle this, or it may be simpler to add a listener for the delete event directly to the Session Resource. There's an example of this for the Task Resource: https://github.com/learningequality/studio/blob/develop/contentcuration/contentcuration/frontend/shared/data/resources.js#L1410

MisRob replied:
Oh, that's nice. I was not too fond of that place either.

rtibbles reviewed (commented):
A couple more code comments, have not manually tested just yet!

MisRob commented:
Thank you, @rtibbles. I've just merged develop and pushed all new updates. If this seems to be the final version when it comes to the overall strategy, I am happy to add some tests before you check it out manually.

MisRob commented:
I've just pushed tests so everything should be in place now.

rtibbles commented on contentcuration/contentcuration/frontend/shared/vuex/session/index.js:
```diff
@@ -109,10 +103,9 @@ export default {
-      });
+    logout() {
+      resetDB();
+      return client.get(window.Urls.logout());
```
Maybe this should be:
```
return client.get(window.Urls.logout()).then(resetDB);
```

As, in combination with the delete listener, we probably want to make sure that the logout occurs server side before we clear the DB and initiate the redirect?

rtibbles reviewed (commented):
One thing that might need to be tweaked. The other thing that occurs to me is that we might be better to edit the `resetDB` function also: https://github.com/learningequality/studio/blob/develop/contentcuration/contentcuration/frontend/shared/data/index.js#L43 to wrap everything in a single transaction, rather than issuing a separate transaction for each table.

```
export function resetDB() {
  const tableNames = Object.values(TABLE_NAMES);
  return db.transaction('rw', ...tableNames, () => {
    Dexie.currentTransaction.source = IGNORED_SOURCE;
    return Promise.all(
      tableNames.map(table => {
        if (db[table]) {         
          return db[table].clear();
        }
        return Promise.resolve();
      });
    });
}
```
I think this will make the reset atomic, and so should prevent the change events being issued while the reset is still in process.

rtibbles commented:
Manual testing all checks out - so it's possible that my concerns are overblown, but I'd rather be safe than sorry!

MisRob commented:
Thank you @rtibbles, both notes make sense to me

MisRob commented:
I've just pushed all updates and also added some adjustments related to using one transaction. Valid tables need to be provided when defining its scope, otherwise if fails on:

![Selection_003](https://user-images.githubusercontent.com/13509191/99389864-fa754080-28d7-11eb-91be-4bb9c50642aa.png)
![Selection_004](https://user-images.githubusercontent.com/13509191/99389867-fb0dd700-28d7-11eb-8c73-ce764317b43a.png)

rtibbles commented:
Ah, looks like we should delete the tree table from our constants too!

rtibbles reviewed (approved):
Changes look good, and final manual testing pass confirms it is still working like a charm. Great work!
